devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+# Cancellation Token Requirements
```

Comment:
>Cancellation Token Requirements [](http://example.com/codeflow?start=2&length=31)

what's the relation between this and ASYNC_OP in c_util?

My understanding is that to cancel something, it needs to be ASYNC ("in general").

So would be really sweet if there would be some clarification around that...

Reply:
Good point. Added the following:

----

### How does this relate to `async_op`?

The `c-util` repo includes a module called [async_op](async_op.md) which provides the following abilities:

- Provide some space on the heap for storing arbitrary data as it pertains to an async operation.
- Allow the calling of a cancel function via the `async_op_cancel` API on the `async_op` handle.

There is some overlap in functionality offered by `async_op` and cancellation tokens. Both can be used to implement cancellation on asynchronous functions. `async_op` can however be used to have only one cancellation callback associated with it. Consider a system for example where an incoming call is serviced by multiple software layers potentially involving multiple I/O calls. In such a system, it might be desirable to have a single cancellation signal trigger a collection of callbacks in various parts of the system that take separate action to implement cancellation. While it is technically feasible to implement such a system with `async_op`, it will involve fairly intricate state management and likely cause tight coupling between the part of the code that does the signalling with the part of the code that implements cancellation.

With cancellation tokens, the cite where the signal is raised does not have any knowledge of what action is taken at the place where the signal is handled. Further, unlike `async_op`, cancellation tokens also support the ability to imperatively check if the signal has been raised and take reactive action as can be necessary in situations where the work being cancelled is CPU-bound instead of being I/O bound.